text/0000-pod.md

```diff
@@ -0,0 +1,138 @@
+
+# Motivation
+
+### `Range` is not `Copy` (or deriving `Copy` is hard)
+
+We have a guideline that says that iterators should not be implicitly copyable
+(i.e. they shouldn't implement `Copy`). The `Range` struct, which is what
+`a..b` desugars to, is an iterator, and under this guideline it shouldn't be
+`Copy`, but this hurts ergonomics in other cases where it's not used as an
+iterator, for example a struct that contains `Range` can't be marked as `Copy`,
+because `Range` is not `Copy`.
+
+This is a more general problem, if upstream decides that some type shouldn't be
+`Copy`, even if it could, then downstream can't make `Copy`able structs/enums
+that contain said type.
+
+### `Cell` only admits `Copy` data
+
+This is unnecessarily restrictive because the data doesn't need to be
+implicitly copyable, the actual requirement is that the data should be
+cloneable by just `memcpy`ing it.
+
+### How splitting `Copy` helps with this?
+
+Implementing `Pod` for a type is much less controversial than implementing
+`Copy` on it. This should result in more types marked as `Pod` in libraries,
+which in turn gives downstream users more control about what can be `Copy` as
+the only requirement is that the fields must be `Pod`.
+
+`Cell` can be modified to have a `Pod` bound instead of a `Copy` one, this
+would allow more types, like iterators, to be stored in a `Cell`.
+
+# Detailed design
+
+### The traits
+
+A `Pod` marker trait will be added to the stdlib, and the `Copy` trait will be
+updated as follows:
+
+```
+/// Indicates that this type can be cloned by just copying its bits (`memcpy`)
+#[lang = "pod"]
+trait Pod: MarkerTrait {}
+
+/// Indicates that this type will be implicitly copied instead of moved
+#[lang = "copy"]
+trait Copy: Pod {}
+```
+
+### Compiler rules
+
+Update the compiler to enforce the following rules:
+
+- `Pod` can only be implemented by structs/enums whose fields are also `Pod`.
+- Built-in/primitive types that were automatically considered `Copy` by the
+  compiler (`i8`, `&T`, `(A, B)`, `[T; N]`, etc), will now be considered to be
+  both `Pod` and `Copy`.
+
+### `#[derive]` syntax extensions
+
+To avoid breaking the world, the `#[derive(Copy)]` syntax extension will be
+updated to derive *both* `Copy` and `Pod`:
+
+``` rust
+#[derive(Copy)]
+struct Foo<T>(T);
+
+// expands into
+impl<T: Pod> Pod for Foo<T> {}
+impl<T: Pod> Copy for Foo<T> {}
+//^ note that the bound is `T: Pod` (not `T: Copy`!)
+```
+
+Also add a `#[derive(Pod)]` extension to derive only `Pod`.
+
+### `Cell`
+
+`Cell` will be change its `T: Copy` bound to `T: Pod`.
+
+# Drawbacks
+
+Cognitive load. Currently you have to think whether a type should be `Clone`,
+`Copy+Clone` or neither (affine type). Now you have to choose between `Clone`,
+`Pod+Clone`, `Pod+Clone+Copy` or neither. (There are other combinations like
+e.g `Pod+Copy`, but you almost always want to add `Clone` in there as well)
```

Comment:
I think there's a general consensus that `Copy` should imply `Clone` unconditionally. Not doing this should be considered a bug IMO. Since this is introducing a multi-derive, might we consider adding more? `#[derive(Copy)]` does `Copy + Pod + Clone`, `#[derive(Ord)]` does `Ord + Eq + PartialEq + PartialOrd`, etc?